gitlab/provider.go

```diff
@@ -118,12 +118,13 @@ func init() {
 
 func providerConfigure(d *schema.ResourceData) (interface{}, error) {
 	config := Config{
-		Token:      d.Get("token").(string),
-		BaseURL:    d.Get("base_url").(string),
-		CACertFile: d.Get("cacert_file").(string),
-		Insecure:   d.Get("insecure").(bool),
-		ClientCert: d.Get("client_cert").(string),
-		ClientKey:  d.Get("client_key").(string),
+		Token:            d.Get("token").(string),
+		BaseURL:          d.Get("base_url").(string),
+		CACertFile:       d.Get("cacert_file").(string),
+		Insecure:         d.Get("insecure").(bool),
+		ClientCert:       d.Get("client_cert").(string),
+		ClientKey:        d.Get("client_key").(string),
+		TerraformVersion: "TODO", // TODO: How to get Terraform version?
```

Comment:
Any advice on how to get the Terraform version? 

I saw `var Version = "0.12.7"` from https://github.com/hashicorp/terraform-plugin-sdk/blob/v1.15.0/internal/version/version.go#L14 but I dont think that is what we want because its an internal module, doesnt exist in v2 of the SDK, and its the version of Terraform the provider is compiled with rather than the version of Terraform the consumer of the provider is running.

I see there is a `TerraformVersion string` attribute in the `schema.Provider struct` at https://github.com/hashicorp/terraform-plugin-sdk/blob/v1.15.0/helper/schema/provider.go#L72 which you instantiate here https://github.com/gitlabhq/terraform-provider-gitlab/blob/master/gitlab/provider.go#L15 but I cant figure out where that value is getting set.

The `State struct` from https://github.com/hashicorp/terraform-plugin-sdk/blob/v1.15.0/terraform/state.go#L87 has a `TFVersion string` attribute but that is just the version of Terraform that wrote the state file. Also, I couldnt figure out how to get the `state struct` instance from `d *schema.ResourceData`.

Is there a way of getting the Terraform version from `d *schema.ResourceData`?

Reply:
> For the user agent, can we use https://github.com/hashicorp/terraform-plugin-sdk/blob/9da7cbac0dfcd4c7096105eeda94ea69f53c1101/helper/schema/provider.go#L451 ?
> 
> Thanks!

Just saw your comment @roidelapluie. I saw that `UserAgent`  function but that function is attached to the `Provider` struct. The functions `providerConfigure` and `config.Client()` dont have access to the `Provider` struct that gets instantiated at https://github.com/gitlabhq/terraform-provider-gitlab/blob/master/gitlab/provider.go#L15

I we have is `d *schema.ResourceData` which from what I can tell doesnt hold a reference to the `provider`.

Reply:
OK, here is an idea of how to get the `provider` struct into the `providerConfigure` function: https://github.com/bishtawi/terraform-provider-gitlab/commit/9fcf005a0fe275e0f67765b8d71cc937017382b4

This idea would work provided that Terraform sets the `provider.TerraformVersion` variable before calling the `ConfigureFunc`. If it sets it after, then `provider.TerraformVersion` will be empty string.

Reply:
Looked at the source of the SDK, looks like `provider.TerraformVersion` is set before `ConfigureFunc` is called.